Add peripheral

trarbr opened the pull request:
This PR adds the base functionality for BLE Peripherals:

- A `BlueHeron.Peripheral` module (and `:gen_statem` process), which is the starting point for library users. It includes:
  - An API for configuring, starting and stopping advertising
  - Handling of received ATT packets (by forwarding them to `BlueHeron. GATT.Server` and then replying to the client device)
- A `BlueHeron. GATT.Server` module, which handles the generic GATT pieces, and specifies a behaviour for the non-generic pieces: 
  - Implements (partial) service discovery for the profile specified by the behaviour implementation.
  - Ensures responses fit inside the ATT MTU
  - Handles reading/writing of long values without involving the library user
- `BlueHeron. GATT.Characteristic` and `BlueHeron. GATT.Service` modules used by the library user to describe the profile of the device.

At this point, the code is functional for a limited set of use cases. A non-comprehensive list of missing features (which I don't find necessary at this point):
- Handling of scan requests (for advertising)
- Authentication (bonding) and encryption (both of which are part of Security Manager Protocol)
- ATT MTU exchange
- Included & secondary services
- Characteristic descriptors and permissions
- GATT client API implementation for the Peripheral

I would like to write some bare-bones docs next. I expect to have them up before the end of the weekend.

fhunleth commented:
@trarbr This so cool! I haven't been using BLE for a while, but I might have to find a project. This looks like a huge improvement.

trarbr commented:
I ended up writing a usage example in the form of a LiveBook instead of module docs 😆  LiveBook on Nerves is just so fun.

https://github.com/trarbr/nerves_livebook/blob/add-ble-peripheral-example/priv/samples/bluetooth/ble_device_with_nerves.livemd

Please try it out, I'll appreciate any feedback 

Module docs will come some time during the week.

fhunleth commented:
This is super cool!!!!

I'm trying it out on a RPi Zero which I don't think is the issue, but I get this message when I try to connect to MyApp-XY in nRF Connect:

<img width="747" alt="image" src="https://user-images.githubusercontent.com/64669/133943013-3982f9d8-d885-4c75-9b53-a4a4ee214bbe.png">

This seems like it might be an easy fix, but I didn't debug.

trarbr commented:
Thank you for taking it for a spin 👍  

Looks like the Peripheral is trying to handle a packet that is not an ATT message (L2Cap cid should be 0x04 for ATT messages on BLE). Also, the ACL handle looks weird - 576? The connection handle is supposed to be 64 according to the process state. It gets this handle from the HCI LE ConnectionComplete event, and that handle is supposed to be used for all the ACL messages. 

I guess I can filter out the packet by being more specific in my match clause (https://github.com/blue-heron/blue_heron/pull/70/files#diff-0de9220626896e4cdc65d6f30409a11d58bfee5bb3b48de2a856599de83f82abR116), but I would love to understand why your device got that packet. Any chance you can share the HCI packet log in `/tmp/hcidump.pklg`?

Also, what phone/tablet did you try nRF Connect on? I've only tried it on an Android phone so far.

fhunleth commented:
This was an iPhone with iOS 14.8. I'll see if I can get a packet capture later tonight.

fhunleth commented:
I tried it again with `97deccc`. It doesn't crash, but it doesn't connect either. Here are the logs:

```
19:13:20.683 [info]  Unhandled ACL packet: %BlueHeron.ACL{data: %BlueHeron.L2Cap{cid: 58, data: <<9, 5, 10, 35, 21, 0, 0>>}, flags: %{bc: 0, pb: 0}, handle: 576}

19:13:20.695 [info]  Unhandled ACL packet: %BlueHeron.ACL{data: %BlueHeron.L2Cap{cid: 4, data
: %BlueHeron.ATT.ExchangeMTURequest{client_rx_mtu: 185, opcode: 2}}, flags: %{bc: 0, pb: 0}, handle: 576}
```

Here's the hcidump.pklg:

[hcidump.pklg.gz](https://github.com/blue-heron/blue_heron/files/7213253/hcidump.pklg.gz)

trarbr commented:
Thank you so much for the packet log, that was really helpful.

I found a bug in the ACL deserialisation. This is why we saw the odd ACL packet handle of 576, which didn't match the handle that we got from the HCI LE Connection Complete event. It's fixed in #72, and I have rebased this branch on that - so the match clause should work now.

I also found an iPad to test with, and found that the `nRF Connect for Mobile` uses a couple GATT procedures that were not used by the `nRF Connect` Android app - so I have implemented those.  I tested both on Android and the iPad, so I hope it works for you now 🤞 

I have updated my Livebook repo to use the latest commit on this branch.

fhunleth commented:
It works for me!!!!!!!

Great work on this!

ConnorRigby reviewed (approved):
Sorry for taking so long on this. It looks really cool to me!
